File: app/src/main/java/xyz/rodit/snapmod/UriResolverSubscriber.java
package xyz.rodit.snapmod;

import android.util.Log;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

import de.robv.android.xposed.XposedBridge;
import xyz.rodit.snapmod.mappings.RxObserver;

public class UriResolverSubscriber implements InvocationHandler {

    private static final String TAG = "UriResolverSubscriber";

    private final ResolutionListener listener;

    public UriResolverSubscriber(ResolutionListener listener) {
        this.listener = listener;
    }

    @Override
    public Object invoke(Object proxy, Method method, Object[] args) {
        if (method.getName().equals(RxObserver.accept.getDexName())) {
            Log.d(TAG, "Resolved content uri - notifying listener.");
            if (listener != null) {
                listener.accept(args[0]);
            }
        } else if (method.getName().equals(RxObserver.error.getDexName())) {
            XposedBridge.log("Error while resolving uri.");
            XposedBridge.log((Throwable) args[0]);
        }

        return null;
    }

    public interface ResolutionListener {

        void accept(Object result);
    }
}

File: app/src/main/java/xyz/rodit/snapmod/UriResolverSubscriber.kt
package xyz.rodit.snapmod

import android.util.Log
import xyz.rodit.snapmod.logging.log
import xyz.rodit.snapmod.mappings.RxObserver
import xyz.rodit.snapmod.util.TAG
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

typealias ResolutionListener = (result: Any?) -> Unit

open class UriResolverSubscriber(private val listener: ResolutionListener) : InvocationHandler {

    override fun invoke(proxy: Any, method: Method, args: Array<Any>): Any? {
        if (method.name == RxObserver.accept.dexName) {
            Log.d(TAG, "Resolved content uri - notifying listener.")
            listener.invoke(args[0])
        } else if (method.name == RxObserver.error.dexName) {
            log.error("Error resolving media uri.", args[0] as Throwable)
        }

        return null
    }
}